src/System.Management.Automation/engine/AsyncByteStreamDrainer.cs

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#nullable enable
+
+using System.Buffers;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Management.Automation;
+
+internal sealed class AsyncByteStreamDrainer : IDisposable
+{
+    private const int DefaultBufferSize = 1024;
+
+    private readonly SpanAction<byte, object?> _callback;
+
+    private readonly Action _callbackCompleted;
+
+    private readonly object? _callbackArg;
+
+    private readonly IStreamSource _streamSource;
+
+    private readonly Memory<byte> _buffer;
+
+    private readonly CancellationTokenSource _cts = new();
+
+    private Task? _readToBufferTask;
+
+    public AsyncByteStreamDrainer(IStreamSource streamSource, SpanAction<byte, object?> callback, object? callbackArg, Action completedCallback)
```

Comment:
So I tried to work in `CopyToAsync`. It would work, *but* both `Process` instances would need to be made before I can run that. That means I'd need to call `NativeCommandProcessor.Prepare` in the opposite order for native commands (e.g. upstream *first*) which in turn makes it more challenging to revert behavior when merging stderr to stdout. I think the logic is simpler when we don't need the `Process` instance created until we actually get data to write downstream

Reply:
`stdOutDrainer.BeginReadChunks()` is called in `InitNativeProcess` which happens in `Prepare`. Doesn't this mean that the downstream native process's input stream needs to be already opened when the current native command processor is still running `Prepare`?

As of today, the downstream native command's `_nativeProcess` is not yet created when the `BeginReadChunks()` from the upstream native command starts to run, so there seems to be a race condition: `_callback` gets called within `BeginReadChunks()`, which calls `_nativeProcess.StandardInput.BaseStream` on the downstream native command processor while `_nativeProcess` is still `null`.